### DISCUSSION TITLE
Returning a vector of pointers managed on the `C++` side

### DISCUSSION THREAD

**Discussion by thelfer** [wjakob/nanobind#301] [Q&A] [answered]
Dear `nanobind` community,

My question is about wrapping a method returning a vector of pointer. The pointers are owned by the `C++` objects.

To illustrate the point, I have something like:

```cxx
struct Element;

struct ElementsManager{
std::vector<Element *>  select(std::string view);
private:
std::vector<std::unique_ptr<Element>> elements;
};
```

How can I wrap the `select` method ? Of course, the better would be to use the automatic type conversion provided by the type casters given in `nanobind/stl/vector.h` so that python would see a `list` of `Element`s.

Thanks for any help.

**Comment by thelfer** [comment]
up ?

**Reply by wjakob** [reply, marked as answer]
I would say: use the normal `std::vector` type caster, but bind the method with a return value policy of `reference_internal`.

(You'll want to read up on RVPs in the documentation if this doesn't make sense)

**Reply by thelfer** [reply]
Thank for the answer. So the idea would be to simply:

```c++
nb::class_<ElementsManager>(m, "ElementsManager")
  .def("select", &ElementsManager::select, nb::rv_policy::reference_internal);
```

The RVP would apply to all elements of the vector ? That's perfect.

Maybe you would be interested by a small comment from a new comer to `nanobind`.

I did not find this behavior described in the documentation (of course I may have missed it in some advanced pages, which I also read, but there is nothing about it in the two pages cited below).

This behavior is a bit counter-intuitive at first:  I would expect the RVP to apply on the returned vector, not to the hold elements.

 Maybe a small description derived from this example in the RVP section (https://nanobind.readthedocs.io/en/latest/ownership.html) and/or in the section relative to the `type_caster` (https://nanobind.readthedocs.io/en/latest/exchanging.html) would be beneficial.
 
 Of course, that's only a suggestion.
